echopop/live/live_biology.py

```diff
@@ -0,0 +1,1377 @@
+
+    # Create dataframe copy
+    data_copy = biology_data.copy()
+
+    # Iterate through dictionary to apply filters (if present)
+    for column, value in filter_dict.items():
+        if column in data_copy.columns:
+            data_copy = data_copy[data_copy[column] == value]
+
+    # Return output
+    return data_copy
+
+
+def merge_trawl_info(biology_dict: dict):
+
+    # Get the trawl information dictionary
+    trawl_info_df = biology_dict["trawl_info_df"]
+
+    # Update `catch_df`
+    biology_dict["catch_df"] = biology_dict["catch_df"].merge(trawl_info_df)
+
+    # Update `length_df`
+    biology_dict["length_df"] = biology_dict["length_df"].merge(trawl_info_df)
+
+    # Update `specimen_df`
+    biology_dict["specimen_df"] = biology_dict["specimen_df"].merge(trawl_info_df)
+
+    # Drop the trawl information
+    del biology_dict["trawl_info_df"]
+
+
+def prepare_length_distribution(file_configuration: dict):
+
+    # Get the length distribution parameters
+    distrib_params = file_configuration["biology"]["length_distribution"]["bins"]
+
+    # Create histogram bins
+    length_bins = np.linspace(
+        **{key: value for key, value in zip(["start", "stop", "num"], distrib_params)}, dtype=float
+    )
+
+    # Get the binwidths
+    binwidth = np.diff(length_bins / 2.0).mean()
+
+    # Generate the equivalent interval boundaries for each bin
+    intervals = np.concatenate([length_bins[:1] - binwidth, length_bins + binwidth])
+
+    # Format as a DataFrame and return the output
+    # ---- Add Categorical interval column
+    length_bins_df = pd.DataFrame(
+        {"length_bin": length_bins, "interval": pd.cut(length_bins, intervals)}
+    )
+    # ---- Add numeric lower boundary
+    length_bins_df["lower"] = length_bins_df["interval"].apply(lambda x: x.left).astype(float)
+    # ---- Add numeric upper boundary
+    length_bins_df["upper"] = length_bins_df["interval"].apply(lambda x: x.right).astype(float)
+
+    # Return the dataframe that will be incorporated into the biological data attribute
+    return length_bins_df
+
+
+def preprocess_biology_data(biology_output: dict, spatial_dict: dict, file_configuration: dict):
+
+    # Get SQL database file
+    biology_db = file_configuration["database"]["biology"]
+
+    # Get contrasts used for filtering the dataset
+    # ---- Species
+    species_filter = file_configuration["species"]["number_code"]
+    # ---- Trawl partition information
+    trawl_filter = file_configuration["biology"]["catch"]["partition"]
+    # ---- Create filter dictionary
+    filter_dict = dict(species_id=species_filter, trawl_partition=trawl_filter)
+
+    # Apply the filter
+    filtered_biology_output = {
+        key: biology_data_filter(df, filter_dict)
+        for key, df in biology_output.items()
+        if isinstance(df, pd.DataFrame) and not df.empty
+    }
+    # ---- Create new data flag
+    file_configuration["length_distribution"] = prepare_length_distribution(file_configuration)
```

Comment:
This addition of "length_distribution" in `self.config` is probably not a good idea because you already have `self.config["biology"]["length_distribution"]`. Since the config dict is defined upfront, I would suggest against adding undefined keys into it in other parts of the code. I think I made another comment also on this.